### PR TITLE
Added user followers/following feature

### PR DESCRIPTION
**Pull request by KimLinhVu**
What
===
* allow user to follow other users
* added a follower and following page to dashboard

Why
===
* save profiles users want to follow to later click on
* allow private users to later only display profile to user they are following

Who
===
@makushline @qibolun 

Test Plan
===
Screen Shot 2022-07-22 at 4 34 29 PM

### CODE REVIEW

File: client/src/components/Dashboard/Dashboard.css [hunk 1]
```diff
@@ -23,24 +22,24 @@
...
+.dashboard .favorites-tab, .followers-tab {
   width: 100%;
   padding: 10px 10px;
   display: none;
 }
 
 .dashboard .carousel {
   width: 100%;
-  background: url('../../../img/lily.jpeg');
+  background: url('/Users/kimlinh/CodePath/Project/Capstone-Project/client/img/lily.jpeg');
```
Comment (makushline) [inline client/src/components/Dashboard/Dashboard.css]: Use relative paths for imports.

File: client/src/components/Dashboard/Dashboard.css [hunk 2]
```diff
@@ -23,24 +22,24 @@
...
   background-position: 5rem;
   height: 35vh;
 }
 
 /* navlinks section */
 .dashboard .nav {
-  margin: 10px 0px;
-  padding: 10px 0px 0px 0px;
+  margin: 10px 0;
+  padding: 10px 0 0 0;
```
Comment (makushline) [inline client/src/components/Dashboard/Dashboard.css]: Add px to all values. There are a few other places in this diff as well that should have px added.
https://www.w3schools.com/css/css_padding.asp

File: client/src/components/Dashboard/Dashboard.jsx [hunk 3]
```diff
@@ -82,20 +93,38 @@ function Dashboard() {
...
+                setFollowingShow(false)
               }} className={`${playlistShow ? 'tab-show' : ''}`}>Playlists</button>
               <button onClick={() =>{
                 setPlaylistShow(false)
                 setFavoriteShow(true)
+                setFollowersShow(false)
+                setFollowingShow(false)
               }} className={`${favoriteShow ? 'tab-show' : ''}`}>Favorites</button>
-              <button>Followers</button>
+              <button onClick={() => {
```
Comment (makushline) [inline client/src/components/Dashboard/Dashboard.jsx:119]: These buttons seem to be getting cumersome. Is it possible to create subclasses of buttons that have the defaults you need? 🤔

File: client/src/components/ProfileCard/ProfileCard.jsx [hunk 4]
```diff
@@ -26,19 +28,24 @@ function ProfileCard({
     setAnchorEl(null);
   };
 
-  useEffect(() => {
-    const fetchUserProfile = async () => {
-      const { data } = await getUserProfile()
-      setProfile(data)
+  const handleImageUpload = (e) => {
```
Comment (makushline) [inline client/src/components/ProfileCard/ProfileCard.jsx]: Where is this called? 🤔

File: client/src/components/FollowersView/FollowersView.jsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+}) {
+  const users = followers ? profile.followers : profile.following
+  return (
+    <div className="followers-view">
+      <div className="playlist-container">
+        <div className="header">
+          <h3>{followers ? `${profile.username}'s Followers` : 'Following'}</h3>
+        </div>
+        <div className="playlists">
+          {users.length !== 0 ? (
```
Comment (makushline) [inline client/src/components/FollowersView/FollowersView.jsx]: I think you're overusing the ternary operator a bit. I recommend really only using it for single line statements and then using if/else for multi-line statements.

File: client/src/utils/users.js [hunk 6]
```diff
@@ -31,4 +31,53 @@ export const getUserProfile = () => {
       "x-access-token": localStorage.getItem('token')
     }
   })
+}
+
+export const getUserProfileById = (userId) => {
+  return axios.get('http://localhost:8888/users/profile-id', {
```
Comment (makushline) [inline client/src/utils/users.js]: You shouldn't commit localhost based paths into main. These should probably be relative. Something like

```javascript
axios.get('/users/profile-id', {
```

File: server/routes/users.js [hunk 7]
```diff
@@ -42,7 +52,70 @@ router.get('/location', jwt.verifyJWT, async (req, res, next) => {
...
+    const userId = req.userId
+    const { otherUserId } = req.body
+    await Users.findOneAndUpdate({ _id: otherUserId }, {
+      $push: {
+        followers: { userId }
+      }
+    })
+    res.status(200).json()
+  } catch (error) {
+
```
Comment (makushline) [inline server/routes/users.js]: Empty catch

**Review by makushline — CHANGES_REQUESTED** [review]

File: client/src/components/FollowersView/FollowersView.jsx (hunk 5 above)
Comment (KimLinhVu) [reply client/src/components/FollowersView/FollowersView.jsx]: I don't think I can use if/else statements in jsx. 
https://react-cn.github.io/react/tips/if-else-in-JSX.html#:~:text=if%2Delse%20statements%20don't,function%20calls%20and%20object%20construction.

File: client/src/components/ProfileCard/ProfileCard.jsx (hunk 4 above)
Comment (KimLinhVu) [reply client/src/components/ProfileCard/ProfileCard.jsx]: Oops, I forgot to delete this. I was trying to have an option to upload an image but I didn't finish it.

File: client/src/components/UserCard/UserCard.jsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      const { data } = await getUserProfileById(userId)
+      setProfile(data)
+    }
+    getUserInfo()
+  }, [])
+  
+  return (
+    <div className='user-card'>
+      {profile && (
+        <>
```
Comment (KimLinhVu) [inline client/src/components/UserCard/UserCard.jsx]: div

File: client/src/components/UserCard/UserCard.jsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+          <img src=''/>
```
Comment (KimLinhVu) [inline client/src/components/UserCard/UserCard.jsx]: placeolder

File: client/src/components/UserPlaylist/UserPlaylist.jsx [hunk 10]
```diff
@@ -1,27 +1,60 @@
...
+    await removeUserFollowing(user._id)
+    await removeUserFollower(user._id)
+    setIsFollowing(!isFollowing)
+  }
 
   return (
     <div className='user-playlist'>
       <h2>{user.username}</h2>
-      {/* <p>{user.location.formatted_address}</p> */}
+      {isFollowing ? <button className='unfollow' onClick={handleOnClickUnfollow}>Unfollow User</button> : <button className='follow' onClick={handleOnClickFollow}>Follow User</button>}
```
Comment (KimLinhVu) [inline client/src/components/UserPlaylist/UserPlaylist.jsx]: if/else

File: client/src/utils/playlist.js [hunk 11]
```diff
@@ -1,7 +1,9 @@
 import axios from 'axios'
 
+axios.defaults.baseURL = 'http://localhost:8888'
```
Comment (KimLinhVu) [inline client/src/utils/playlist.js]: move to defaults

File: client/src/utils/spotify.js [hunk 12]
```diff
@@ -88,39 +88,40 @@ const getAccessToken = () => {
 
 export const accessToken = getAccessToken()
 
-/* axios global request headers */
-axios.defaults.baseURL = 'https://api.spotify.com/v1'
-axios.defaults.headers['Authorization'] = `Bearer ${accessToken}`
-axios.defaults.headers['Content-Type'] = 'application/json'
+const instance = axios.create({
```
Comment (KimLinhVu) [inline client/src/utils/spotify.js:91]: defaults

**Review by makushline — APPROVED** [review]

**Pull request merged**